pkg/exemplar/interface.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exemplar
+
+import (
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// TsExemplar is an exemplar alongside a timestamp.
+type TsExemplar struct {
+	Ts       int64
+	Exemplar Exemplar
+}
+
+// SeriesExemplar is a time series of exemplar values.
+type SeriesExemplar []TsExemplar
+
+// The following are similar to those of `storage/interfaces.go`:
+// https://github.com/prometheus/prometheus/blob/91d7175eaac18b00e370965f3a8186cc40bf9f55/storage/interface.go#L31-L65
+
+// Exemplars is an exemplar storage backend.
+type Exemplars interface {
+	// Get is a fast lookup to find an exemplar with an exact match on labels
+	// and timestamp provided (so this can be O(1) lookup).
+	Get(l labels.Labels, t int64) (Exemplar, bool, error)
+
+	// Query is a query to find an exemplars with partial matches on labels
+	// and timestamp ranges provided (this is a range query). Query returns
+	// a list of SeriesExemplar for each label match.
+	Query(l labels.Labels, start, end int64) ([]SeriesExemplar, error)
```

Comment:
To really implement a generic query API, we needed regexp matchers and negative matchers, too.

That's where I'm thinking we should perhaps stay with the `Get` only approach, just that it takes a time range, not just one timestamp, and returns a time series of matchers for that one exact label set. The implementation of the externally visible API would need to hit the normal storage first to resolve the label matchers into concrete series first. I think that's fine for an MVP.

Reply:
Couldn't we do this via `Query(start, end int64, matchers ...*labels.Matcher)`?

Reply:
Sure, but then we are really talking about implementing a whole query engine.
By doing this via a call to the existing query engine, we would waste a bit of resources (as we only want to get the matching label sets and would never access the actual time series data), but it would be much simpler.

Reply:
Also, we didn't need what I commented above (the separate `struct` to show in the answer which exact label set is attached to each returned series of exemplars).

Reply:
One more round of changes. I got rid of `Query` for the time being as query in a way indicates partial matches. Added a `GetRange` which takes a range and returns a list of exemplars based on exact label matches.